<a href="https://colab.research.google.com/github/HowardHNguyen/Python/blob/main/Stock_Market_Comparison_Analysis_using_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# To collect the data, I’ll be using the Yahoo Finance API. You can install the API by executing the pip command mentioned below on your terminal or command prompt:

In [ ]:
# !pip install yfinance

# Now, let’s import the necessary Python libraries to get started with the task:

In [1]:
import pandas as pd
import yfinance as yf
import plotly.io as pio
import plotly.graph_objects as go
pio.templates.default = "plotly_white"

# Now, let’s collect stock price data:

In [10]:
# Define the tickers for Apple, Google, and UHC
apple_ticker = 'AAPL'
google_ticker = 'GOOGL'
uhc_ticker = 'UNH'
tesla_ticker = 'TSLA'

# Define the date range for the last quarter
start_date = '2023-01-02'
end_date = '2023-11-15'

# Fetch historical stock price data using yfinance
apple_data = yf.download(apple_ticker, start=start_date, end=end_date)
google_data = yf.download(google_ticker, start=start_date, end=end_date)
uhc_data = yf.download(uhc_ticker, start=start_date, end=end_date)
tesla_data = yf.download(tesla_ticker, start=start_date, end=end_date)

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


The above code sets up the analysis by defining the companies and date range of interest, and it fetches the historical stock price data for Apple, Google, UHC, and Tesla within that period. This data will serve as the basis for various financial analyses and comparisons, such as calculating returns, volatility, and other metrics to assess the performance and risk associated with these stocks.

# Now let’s calculate the daily returns of Apple, Google, UHC, and Tesla in the stock market:

In [11]:
# Calculate daily returns
apple_data['Daily_Return'] = apple_data['Adj Close'].pct_change()
google_data['Daily_Return'] = google_data['Adj Close'].pct_change()
uhc_data['Daily_Return'] = uhc_data['Adj Close'].pct_change()
tesla_data['Daily_Return'] = tesla_data['Adj Close'].pct_change()

# Now, let’s visualize the daily returns:

In [12]:
# Create a figure to visualize the daily returns
fig = go.Figure()

fig.add_trace(go.Scatter(x=apple_data.index, y=apple_data['Daily_Return'],
                         mode='lines', name='Apple', line=dict(color='blue')))
fig.add_trace(go.Scatter(x=google_data.index, y=google_data['Daily_Return'],
                         mode='lines', name='Google', line=dict(color='green')))
fig.add_trace(go.Scatter(x=uhc_data.index, y=uhc_data['Daily_Return'],
                         mode='lines', name='UHC', line=dict(color='yellow')))
fig.add_trace(go.Scatter(x=tesla_data.index, y=tesla_data['Daily_Return'],
                         mode='lines', name='Tesla', line=dict(color='red')))

fig.update_layout(title='Daily Returns for Apple, Google, UHC, and Tesla (YTD)',
                  xaxis_title='Date', yaxis_title='Daily Return',
                  legend=dict(x=0.02, y=0.95))

fig.show()

# Now let’s analyze the cumulative returns of Apple, Google, UHC, and Tesla - YTD:

In [16]:
# Calculate cumulative returns for YTD
apple_cumulative_return = (1 + apple_data['Daily_Return']).cumprod() - 1
google_cumulative_return = (1 + google_data['Daily_Return']).cumprod() - 1
uhc_cumulative_return = (1 + uhc_data['Daily_Return']).cumprod() - 1
tesla_cumulative_return = (1 + tesla_data['Daily_Return']).cumprod() - 1

# Create a figure to visualize the cumulative returns
fig = go.Figure()

fig.add_trace(go.Scatter(x=apple_cumulative_return.index, y=apple_cumulative_return,
                         mode='lines', name='Apple', line=dict(color='blue')))
fig.add_trace(go.Scatter(x=google_cumulative_return.index, y=google_cumulative_return,
                         mode='lines', name='Google', line=dict(color='green')))
fig.add_trace(go.Scatter(x=uhc_cumulative_return.index, y=uhc_cumulative_return,
                         mode='lines', name='UHC', line=dict(color='yellow')))
fig.add_trace(go.Scatter(x=tesla_cumulative_return.index, y=tesla_cumulative_return,
                         mode='lines', name='Tesla', line=dict(color='red')))

fig.update_layout(title='Cumulative Returns for Apple, Google, UHC, and Tesla (YTD)',
                  xaxis_title='Date', yaxis_title='Cumulative Return',
                  legend=dict(x=0.02, y=0.95))

fig.show()

Here, we first calculated the cumulative returns for Apple, Google, UHC, and Tesla over YTD. Cumulative returns represent the total percentage change in the stock’s value over a given period, considering the compounding effect of daily returns. Then, we compared the investment performance of Apple, Google, UHC, and Tesla during the specified period, showing which stock had higher or lower cumulative returns over that time frame. We can see that Tesla has highest cumulative returns compared to others over YTD records.

# Now, let’s analyze the volatility of Apple, Google, UHC, and Tesla:

In [17]:
# Calculate historical volatility (standard deviation of daily returns)
apple_volatility = apple_data['Daily_Return'].std()
google_volatility = google_data['Daily_Return'].std()
uhc_volatility = uhc_data['Daily_Return'].std()
tesla_volatility = tesla_data['Daily_Return'].std()

# Create a figure to compare volatility
fig1 = go.Figure()
fig1.add_bar(x=['Apple', 'Google', 'UHC', 'Tesla'], y=[apple_volatility, google_volatility, uhc_volatility, tesla_volatility],
             text=[f'{apple_volatility:.4f}', f'{google_volatility:.4f}', f'{uhc_volatility:.4f}', f'{tesla_volatility:.4f}'],
             textposition='auto', marker=dict(color=['blue', 'green', 'yellow', 'red']))

fig1.update_layout(title='Volatility Comparison (YTD)',
                   xaxis_title='Stock', yaxis_title='Volatility (Standard Deviation)',
                   bargap=0.5)
fig1.show()

We first calculated the historical volatility for Apple, Google, UHC, and Tesla stocks. Volatility is a measure of how much the stock’s price fluctuates over time. In this case, we are calculating the standard deviation of daily returns to measure the volatility. Then we visualized the calculated volatility to assess and compare the volatility or risk associated with both Apple, Google, UHC, and Tesla stocks during the specified period. We can see that Tesla’s volatility is higher than others.

It indicates that Tesla’s stock price experienced larger price fluctuations or greater price variability over YTD. Here’s what this difference in volatility may indicate:

 1 - Tesla’s stock is considered riskier compared to others. Investors generally associate higher volatility with higher risk because it implies that the stock price can change significantly in a short period.

 2 - Tesla’s stock may be more sensitive to market conditions, economic factors, or company-specific news and events. This heightened sensitivity can result in larger price swings.

 3 - Traders and investors with a higher risk tolerance might find Tesla’s stock appealing if they are looking for opportunities to profit from short-term price movements.

# Now let’s compare the stock market of Google, Apple, UHC, and Tesla according to the stock market benchmark:

In [19]:
market_data = yf.download('^GSPC', start=start_date, end=end_date)  # S&P 500 index as the market benchmark

# Calculate daily returns for both stocks and the market
apple_data['Daily_Return'] = apple_data['Adj Close'].pct_change()
google_data['Daily_Return'] = google_data['Adj Close'].pct_change()
uhc_data['Daily_Return'] = uhc_data['Adj Close'].pct_change()
tesla_data['Daily_Return'] = tesla_data['Adj Close'].pct_change()

market_data['Daily_Return'] = market_data['Adj Close'].pct_change()

# Calculate Beta for Apple and Google
cov_apple = apple_data['Daily_Return'].cov(market_data['Daily_Return'])
var_market = market_data['Daily_Return'].var()
beta_apple = cov_apple / var_market

cov_google = google_data['Daily_Return'].cov(market_data['Daily_Return'])
beta_google = cov_google / var_market

cov_uhc = uhc_data['Daily_Return'].cov(market_data['Daily_Return'])
beta_uhc = cov_uhc / var_market

cov_tesla = tesla_data['Daily_Return'].cov(market_data['Daily_Return'])
beta_tesla = cov_tesla / var_market

# Compare Beta values
#if beta_apple > beta_google:
#    conclusion = "Apple is more volatile (higher Beta) compared to Google."
#else:
#    conclusion = "Google is more volatile (higher Beta) compared to Apple."

# Print the conclusion
print("Beta for Apple:", beta_apple)
print("Beta for Google:", beta_google)
print("Beta for UHC:", beta_uhc)
print("Beta for Tesla:", beta_tesla)
#print(conclusion)

[*********************100%%**********************]  1 of 1 completed
Beta for Apple: 1.123195016535523
Beta for Google: 1.429394179055039
Beta for UHC: 0.27171118987820014
Beta for Tesla: 2.2653698291347153


In the above code, we are assessing how sensitive Apple, Google, UHC, and Tesla stocks are to overall market movements, providing insights into their relative volatility and risk about the broader U.S. stock market represented by the S&P 500 index.

The Standard & Poor’s 500, often referred to as the S&P 500, is a widely recognized stock market index in the United States. The S&P 500 index includes 500 of the largest publicly traded companies in the United States, chosen for their market capitalization, liquidity, and industry representation. These companies span various sectors of the U.S. economy and provide a comprehensive view of the health and performance of the stock market.

In the above output, the beta value for Apple is approximately 1.123. This beta value suggests that Apple’s stock is estimated to be approximately 22.57% more volatile or sensitive to market movements (as represented by the S&P 500 index) compared to the overall market. The beta value for Google is approximately 1.429. This beta value suggests that Google’s stock is estimated to be approximately 53.03% more volatile or sensitive to market movements. In addition, the beta value for Tesla's stock is approximately 2.265. This beta value suggests that Tesla’s stock is estimated to be approximately 65% more volatile or sensitive to market movements.

A beta greater than 1 suggests that a stock tends to be more volatile than the market. In this case, both Apple and Google have beta values greater than 1, indicating that they are expected to be more volatile and sensitive to market movements.
Tesla's highest beta value compared to Google’s higher beta value (1.5303) compared to Apple’s (1.2257) and UHC's (0.271) suggests that Tesla’s stock is estimated to have a higher degree of market sensitivity or risk compared to others. Investors should consider this information when making investment decisions, as higher-beta stocks can provide greater potential for returns but also carry a higher level of risk.